### ISSUE TITLE
Feature request: Support macOS Backup Core excludes

### ISSUE THREAD

**Issue by da2x** [restic/restic#2180]
Output of `restic version`
--------------------------
restic 0.9.4 compiled with go1.11.4 on darwin/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
Check files to see if they’ve been excluded from inclusion in user backups in Backup Core (Time Machine). There is an API at [CSBackupIsItemExcluded](https://developer.apple.com/documentation/coreservices/1443602-csbackupisitemexcluded).

macOS app developers can use Backup Core to exclude files from backups. Usually non-essential files, caches, and the like. Technically this sets an extended file system attribute called `com.apple.metadata:com_apple_backup_excludeItem`.

Users can manually mark files and directories from exclusion from the Time Machine UI. This doesn’t set an extended attribute on the file. However, the above API (or `tmutil isexcluded`) will correctly return a file’s backup inclusion status.

What are you trying to do?
--------------------------
Avoid backing up ephemeral files by relying on app developers who know their apps’ internal well and know what should be excluded from backups.

Did restic help you or made you happy in any way?
-------------------------------------------------
Deduplicated backups is enough to make anyone happy.

**Comment by rawtaz** [comment]
I wonder if this is something that should really be a concern of restic. It's very very platform specific, and can easily be dealt with by a wrapper script. I do it that way (not really the same thing though, I produce an exclude file with in part the output from an `mdfind` command).

**Comment by nicolas17** [comment]
I disagree that this can be "easily" done with a wrapper script; it would involve *two* whole traversals of the filesystem, one from the wrapper script searching for excluded files (and create the exclusion file), and another from restic. If restic itself checked the attribute, it would be done along with its normal file attribute checks in the same traversal.

**Comment by borkd** [comment]
@nicolas17 - if you have a script that produces OSX release specific standard exclusions plus tmutil/mdfind/xattrs findings from live fs, maybe open a PR against https://github.com/restic/restic/tree/master/contrib/generate-excludes-CSBackupIsItemExcluded-darwin-X.Y.sh ?

**Comment by gsnedders** [comment]
This effectively overlaps with #1242, which asks for a generic means to exclude based on an xattr.

**Comment by tisba** [comment]
> I wonder if this is something that should really be a concern of restic. It's very very platform specific, and can easily be dealt with by a wrapper script. I do it that way (not really the same thing though, I produce an exclude file with in part the output from an `mdfind` command).

While it is relatively easy to create a wrapper script (instead of https://github.com/restic/restic/issues/3714) to create an APFS snapshot to get a consistent backup, `mdfind` (e.g. as `mdfind "com_apple_backup_excludeItem = 'com.apple.backupd'"`) will use the spotlight index, which is not guaranteed to be in sync. Relying on it for the purpose of backups is calling for trouble.

(Disclaimer: I'm not a heavy restic user, nor do I know the codebase) Searching the documentation looks like restic does support extended attributes (https://restic.readthedocs.io/en/latest/manual_rest.html#metadata-handling). So at least in theory it should be possible to include/exclude based on them.
